x/poolmanager/client/query_proto_wrap.go

```diff
@@ -254,3 +256,162 @@ func (q Querier) TotalLiquidity(ctx sdk.Context, req queryproto.TotalLiquidityRe
 		Liquidity: totalLiquidity,
 	}, nil
 }
+
+// EstimateTradeBasedOnPriceImpact returns the input and output amount of coins for a pool trade
+// based on twap value and maximum price impact.
+func (q Querier) EstimateTradeBasedOnPriceImpact(
+	ctx sdk.Context,
+	req queryproto.EstimateTradeBasedOnPriceImpactRequest,
+) (*queryproto.EstimateTradeBasedOnPriceImpactResponse, error) {
+	if req.PoolId == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Invalid Pool Id")
+	}
+
+	if req.FromCoin.Denom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid from coin denom")
+	}
+
+	if req.ToCoinDenom == "" {
+		return nil, status.Error(codes.InvalidArgument, "invalid to coin denom")
+	}
+
+	// When calculating the estimate one or more of the queries we are using might
+	// panic, this is problematic when using this query in async-icq.
+	var resp *queryproto.EstimateTradeBasedOnPriceImpactResponse
+	var err error
+	err = osmoutils.ApplyFuncIfNoError(ctx, func(ctx sdk.Context) error {
+		resp, err = q.estimateTradeBasedOnPriceImpactInternal(ctx, req)
+		return err
+	})
```

Comment:
Can you link the async-icq issue/PR that would remove the need for ApplyFuncIfNoError?